crates/data_components/src/duckdb/write.rs

```diff
@@ -187,32 +212,42 @@ impl DuckDBDataSink {
             })?;
         }
 
+        let mut insert_table = orig_table_creator
+            .create_empty_clone()
+            .map_err(to_datafusion_error)?;
+
+        let Some(insert_table_creator) = insert_table.table_creator.take() else {
+            unreachable!()
+        };
+
         for (i, batch) in data_batches.iter().enumerate() {
             tracing::debug!(
                 "Inserting batch #{i}/{} into cloned table.",
                 data_batches.len()
             );
-            insert_table
-                .insert_batch(tx, batch)
-                .map_err(to_datafusion_error)?;
+            if let Err(e) = insert_table
+                .insert_batch_no_constraints(tx, batch)
+                .map_err(to_datafusion_error)
+            {
+                insert_table_creator
+                    .delete_table(tx)
+                    .map_err(to_datafusion_error)?;
+                return Err(e);
+            };
         }
 
-        let Some(insert_table_creator) = insert_table.table_creator.take() else {
-            unreachable!()
-        };
-
         if self.overwrite {
             insert_table_creator
                 .replace_table(tx, orig_table_creator)
                 .map_err(to_datafusion_error)?;
         } else {
-            // Specific on-conflict handling will be done here.
-            insert_table
-                .insert_table_into(tx, &self.duckdb)
-                .map_err(to_datafusion_error)?;
+            let insert_result = insert_table
+                .insert_table_into(tx, &self.duckdb, self.on_conflict.as_ref())
+                .map_err(to_datafusion_error);
```

Comment:
If the data update is an `overwrite` then we just replace the old table with the temp table. But if its an Append, then we go down the path of `insert_table_into` which will take the `on_conflict` behavior into account.